CIP-XXXX/README.md

```diff
@@ -0,0 +1,69 @@
+
+# Specification
+
+All addresses in Cardano currently start with a header that indicates which address type they are. Notably, these can be seen in [CIP19](../CIP-0019/README.md).
+
+Note there are two ways we can support our modification to the address type:
+
+1) Reserve a new address type as part of the header nibble. This is not great because it means we have to reserve many new address types (a new alternative for every case where the old address format contains a script hash)
+2) Add a new column to existing address types that contain scripts. This doesn't work for pointer addresses because they are variable-width so it isn't possible to differentiate between the original address format and this annotated format (note that pointer addresses are almost never used on chain, so this is an acceptable limitation). This does, however, limit the kinds of modifications we could make to these address types in the future as we would need to make sure the length continues to uniquely identify the content.
+
+In this CIP, we propose using option (2) and extend the CIP19 table as follows:
+
+Header type (`t t t t . . . .`) | script namespace* | Payment Part     | script namespace* | Delegation Part
+---                             | ---               | ---              | ---               |---
+(0) `0000....`                  | ø                 | `PaymentKeyHash` | ø                 | `StakeKeyHash`
+(1) `0001....`                  | 0-255             | `ScriptHash`     | ø                 | `StakeKeyHash`
+(2) `0010....`                  | ø                 | `PaymentKeyHash` | 0-255             | `ScriptHash`
+(3) `0011....`                  | 0-255             | `ScriptHash`     | 0-255             | `ScriptHash`
+(6) `0110....`                  | ø                 | `PaymentKeyHash` | ø                 | ø
+(7) `0111....`                  | 0-255             | `ScriptHash`     | ø                 | ø
+
+**script namespace* is defined to be the key used to identify the script kind in the `script` of the [binary spec](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl#L421). Note that this is the same namespace used to generate a script hash from script bytes.
+
+# Asserted language and script language mismatch
+
+Note that nothing stops somebody from creating a mal-constructed address where the language asserted inside the address does not match the actual language used for the script hash.
+
+This is an unlikely error since addresses are provided to the users by dApps. Additionally, even if due to an error the wrong address is used, funds are not permanently lost because the spending condition is still controlled by the script hash -- meaning they can still be spent and additionally can be considered 
```

Comment:
Are you proposing that we do not validate the script namespace at the time of spending, when the scrip type becomes known? I worry about adding in a field that is not enforced. If folks cannot depend in the field being correct, it is dangerous to write code with that assumption (even if it normally hold true).

Reply:
Well the alternative is that if they get the namespace wrong then the funds are locked forever. Neither are great options

That being said, if we prefer to take neither option and avoid implementing this at the protocol level, having this supported as a user-level annotated address is still useful

Reply:
I would prefer this to be checked.

Reply:
some thoughts for how to make these checked: make that any tx that sends to an annotated script address needs to either provide the script in any of:
1. The witness
2. The auxiliary data
3. A reference input
It makes using this feature a bit more expensive on-chain and a bit more tedious for wallets to implement, but allows having the correctness of the transaction be verifiable using the existing transaction context

Any other alternative to make this checked that includes something like "the full scrip for this hash has to appear somewhere on chain before" requires expanding the transaction validation context which may not be desirable.

Reply:
To clarify, I think we could have any or all of the following kinds of checking:
1. Checking upon spending such an output. This should be easy, since the witness is already there.
2. Checking, locally, before creating such an output i.e. in `cardano-api`. This would require providing some extra information at transaction-creation time, but wouldn't bloat the transaction.
3. Checking, on-chain, when creating such an output. This would require including the witness when creating the transaction, which is significantly more expensive *and* is a much bigger privacy leak since the preimage really would be publicly available on-chain before the output was spent.

I'm in favour of 1 (since it's cheap), optionally 2 (if they have the information, they might as well tell us for cross-checking), and probably not 3 (although I guess we could support it optionally?).